### ISSUE TITLE
NodeJS Functions running in VSCode shutting down after 15-20 minutes

### ISSUE THREAD

**Issue by kpm-at-hfi** [Azure/azure-functions-nodejs-worker#351]
I'm not sure this is the right place for this issue, but it's hard to know for sure. Please redirect me if needed.

In VSCode I'm running a NodeJS Functions app. It starts and runs fine for 15-20 minutes, and then I see:

```
Exceeded language worker restart retry count for runtime:node. Shutting down and proactively recycling the Functions Host to recover
```

I don't remember seeing this happening for this project when I was working daily on it a few weeks ago.

Basic stats:

* Node: v10
* Core Tools Version: 3.0.2996 Commit hash: c54cdc36323e9543ba11fb61dd107616e9022bba
* Function Runtime Version: 3.0.14916.0
* Happens on Mac or Windows
* Happens when running locally or when running in a remote container
* While it's running, it functions perfectly
* The project is _very_ simple APIs getting data from a DB
* The project is running in an Azure Functions App fronted by APIM with no (known) problems

`local.settings.json` looks like:
```json
{
    "IsEncrypted": false,
    "Values": {
        "AzureWebJobsStorage": "",
        "FUNCTIONS_WORKER_RUNTIME": "node",
        "NO_PROXY": "localhost,127.0.0.1"
    },
    "Host": {
        "LocalHttpPort": 7071,
        "CORS": "*"
    }
}
```

Running `func start --verbose` looks like:
```
                  %%%%%%
                 %%%%%%
            @   %%%%%%    @
          @@   %%%%%%      @@
       @@@    %%%%%%%%%%%    @@@
     @@      %%%%%%%%%%        @@
       @@         %%%%       @@
         @@      %%%       @@
           @@    %%      @@
                %%
                %


Azure Functions Core Tools
Core Tools Version:       3.0.2996 Commit hash: c54cdc36323e9543ba11fb61dd107616e9022bba
Function Runtime Version: 3.0.14916.0

[2020-11-18T23:05:45.268Z] Building host: startup suppressed: 'False', configuration suppressed: 'False', startup operation id: 'ec0d6403-30cb-44d4-8a80-97977a303e9e'
[2020-11-18T23:05:45.312Z] Reading host configuration file '/workspaces/<project>/host.json'
[2020-11-18T23:05:45.315Z] Host configuration file read:
[2020-11-18T23:05:45.315Z] {
[2020-11-18T23:05:45.315Z]   "version": "2.0",
[2020-11-18T23:05:45.315Z]   "logging": {
[2020-11-18T23:05:45.315Z]     "applicationInsights": {
[2020-11-18T23:05:45.316Z]       "samplingSettings": {
[2020-11-18T23:05:45.316Z]         "isEnabled": true,
[2020-11-18T23:05:45.316Z]         "excludedTypes": "Request"
[2020-11-18T23:05:45.316Z]       }
[2020-11-18T23:05:45.316Z]     }
[2020-11-18T23:05:45.316Z]   },
[2020-11-18T23:05:45.316Z]   "extensionBundle": {
[2020-11-18T23:05:45.316Z]     "id": "Microsoft.Azure.Functions.ExtensionBundle",
[2020-11-18T23:05:45.316Z]     "version": "[1.*, 2.0.0)"
[2020-11-18T23:05:45.316Z]   }
[2020-11-18T23:05:45.316Z] }
[2020-11-18T23:05:45.345Z] Loading functions metadata
[2020-11-18T23:05:45.376Z] FUNCTIONS_WORKER_RUNTIME set to node. Skipping WorkerConfig for language:java
[2020-11-18T23:05:45.379Z] FUNCTIONS_WORKER_RUNTIME set to node. Skipping WorkerConfig for language:powershell
[2020-11-18T23:05:45.384Z] FUNCTIONS_WORKER_RUNTIME set to node. Skipping WorkerConfig for language:python
[2020-11-18T23:05:45.387Z] Reading functions metadata
[2020-11-18T23:05:45.436Z] 10 functions found
[2020-11-18T23:05:45.457Z] 10 functions loaded
[2020-11-18T23:05:45.466Z] Looking for extension bundle Microsoft.Azure.Functions.ExtensionBundle at /home/node/.azure-functions-core-tools/Functions/ExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle
[2020-11-18T23:05:45.471Z] Found a matching extension bundle at /home/node/.azure-functions-core-tools/Functions/ExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/1.3.2
[2020-11-18T23:05:45.474Z] Fetching information on versions of extension bundle Microsoft.Azure.Functions.ExtensionBundle available on https://functionscdn.azureedge.net/public/ExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/index.json
[2020-11-18T23:05:45.647Z] Skipping bundle download since it already exists at path /home/node/.azure-functions-core-tools/Functions/ExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/1.3.2
[2020-11-18T23:05:45.658Z] Loading extension bundle from /home/node/.azure-functions-core-tools/Functions/ExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/1.3.2/bin
[2020-11-18T23:05:45.673Z] Reading host configuration file '/workspaces/<project>/host.json'
[2020-11-18T23:05:45.673Z] Host configuration file read:
[2020-11-18T23:05:45.673Z] {
[2020-11-18T23:05:45.673Z]   "version": "2.0",
[2020-11-18T23:05:45.673Z]   "logging": {
[2020-11-18T23:05:45.673Z]     "applicationInsights": {
[2020-11-18T23:05:45.673Z]       "samplingSettings": {
[2020-11-18T23:05:45.673Z]         "isEnabled": true,
[2020-11-18T23:05:45.673Z]         "excludedTypes": "Request"
[2020-11-18T23:05:45.674Z]       }
[2020-11-18T23:05:45.674Z]     }
[2020-11-18T23:05:45.674Z]   },
[2020-11-18T23:05:45.674Z]   "extensionBundle": {
[2020-11-18T23:05:45.674Z]     "id": "Microsoft.Azure.Functions.ExtensionBundle",
[2020-11-18T23:05:45.674Z]     "version": "[1.*, 2.0.0)"
[2020-11-18T23:05:45.674Z]   }
[2020-11-18T23:05:45.675Z] }
[2020-11-18T23:05:46.117Z] FUNCTIONS_WORKER_RUNTIME set to node. Skipping WorkerConfig for language:java
[2020-11-18T23:05:46.117Z] FUNCTIONS_WORKER_RUNTIME set to node. Skipping WorkerConfig for language:powershell
[2020-11-18T23:05:46.118Z] FUNCTIONS_WORKER_RUNTIME set to node. Skipping WorkerConfig for language:python
[2020-11-18T23:05:46.245Z] Initializing Warmup Extension.
[2020-11-18T23:05:46.317Z] Initializing Host. OperationId: 'ec0d6403-30cb-44d4-8a80-97977a303e9e'.
[2020-11-18T23:05:46.329Z] Host initialization: ConsecutiveErrors=0, StartupCount=1, OperationId=ec0d6403-30cb-44d4-8a80-97977a303e9e
[2020-11-18T23:05:46.369Z] LoggerFilterOptions
[2020-11-18T23:05:46.369Z] {
[2020-11-18T23:05:46.369Z]   "MinLevel": "None",
[2020-11-18T23:05:46.369Z]   "Rules": [
[2020-11-18T23:05:46.370Z]     {
[2020-11-18T23:05:46.370Z]       "ProviderName": null,
[2020-11-18T23:05:46.370Z]       "CategoryName": null,
[2020-11-18T23:05:46.370Z]       "LogLevel": null,
[2020-11-18T23:05:46.370Z]       "Filter": "<AddFilter>b__0"
[2020-11-18T23:05:46.370Z]     },
[2020-11-18T23:05:46.370Z]     {
[2020-11-18T23:05:46.370Z]       "ProviderName": "Microsoft.Azure.WebJobs.Script.WebHost.Diagnostics.SystemLoggerProvider",
[2020-11-18T23:05:46.370Z]       "CategoryName": null,
[2020-11-18T23:05:46.370Z]       "LogLevel": "None",
[2020-11-18T23:05:46.370Z]       "Filter": null
[2020-11-18T23:05:46.370Z]     },
[2020-11-18T23:05:46.370Z]     {
[2020-11-18T23:05:46.371Z]       "ProviderName": "Microsoft.Azure.WebJobs.Script.WebHost.Diagnostics.SystemLoggerProvider",
[2020-11-18T23:05:46.371Z]       "CategoryName": null,
[2020-11-18T23:05:46.371Z]       "LogLevel": null,
[2020-11-18T23:05:46.371Z]       "Filter": "<AddFilter>b__0"
[2020-11-18T23:05:46.371Z]     },
[2020-11-18T23:05:46.371Z]     {
[2020-11-18T23:05:46.371Z]       "ProviderName": "Azure.Functions.Cli.Diagnostics.ColoredConsoleLoggerProvider",
[2020-11-18T23:05:46.371Z]       "CategoryName": null,
[2020-11-18T23:05:46.371Z]       "LogLevel": null,
[2020-11-18T23:05:46.371Z]       "Filter": "<AddFilter>b__0"
[2020-11-18T23:05:46.371Z]     }
[2020-11-18T23:05:46.371Z]   ]
[2020-11-18T23:05:46.371Z] }
[2020-11-18T23:05:46.372Z] LoggerFilterOptions
[2020-11-18T23:05:46.372Z] {
[2020-11-18T23:05:46.372Z]   "MinLevel": "None",
[2020-11-18T23:05:46.372Z]   "Rules": [
[2020-11-18T23:05:46.372Z]     {
[2020-11-18T23:05:46.372Z]       "ProviderName": null,
[2020-11-18T23:05:46.372Z]       "CategoryName": null,
[2020-11-18T23:05:46.372Z]       "LogLevel": null,
[2020-11-18T23:05:46.373Z]       "Filter": "<AddFilter>b__0"
[2020-11-18T23:05:46.373Z]     },
[2020-11-18T23:05:46.373Z]     {
[2020-11-18T23:05:46.373Z]       "ProviderName": "Microsoft.Azure.WebJobs.Script.WebHost.Diagnostics.SystemLoggerProvider",
[2020-11-18T23:05:46.373Z]       "CategoryName": null,
[2020-11-18T23:05:46.373Z]       "LogLevel": "None",
[2020-11-18T23:05:46.373Z]       "Filter": null
[2020-11-18T23:05:46.373Z]     },
[2020-11-18T23:05:46.373Z]     {
[2020-11-18T23:05:46.373Z]       "ProviderName": "Microsoft.Azure.WebJobs.Script.WebHost.Diagnostics.SystemLoggerProvider",
[2020-11-18T23:05:46.373Z]       "CategoryName": null,
[2020-11-18T23:05:46.373Z]       "LogLevel": null,
[2020-11-18T23:05:46.373Z]       "Filter": "<AddFilter>b__0"
[2020-11-18T23:05:46.374Z]     },
[2020-11-18T23:05:46.374Z]     {
[2020-11-18T23:05:46.374Z]       "ProviderName": "Azure.Functions.Cli.Diagnostics.ColoredConsoleLoggerProvider",
[2020-11-18T23:05:46.374Z]       "CategoryName": null,
[2020-11-18T23:05:46.374Z]       "LogLevel": null,
[2020-11-18T23:05:46.374Z]       "Filter": "<AddFilter>b__0"
[2020-11-18T23:05:46.374Z]     }
[2020-11-18T23:05:46.374Z]   ]
[2020-11-18T23:05:46.374Z] }
[2020-11-18T23:05:46.374Z] FunctionResultAggregatorOptions
[2020-11-18T23:05:46.374Z] {
[2020-11-18T23:05:46.375Z]   "BatchSize": 1000,
[2020-11-18T23:05:46.375Z]   "FlushTimeout": "00:00:30",
[2020-11-18T23:05:46.375Z]   "IsEnabled": true
[2020-11-18T23:05:46.375Z] }
[2020-11-18T23:05:46.375Z] SingletonOptions
[2020-11-18T23:05:46.375Z] {
[2020-11-18T23:05:46.375Z]   "LockPeriod": "00:00:15",
[2020-11-18T23:05:46.375Z]   "ListenerLockPeriod": "00:00:15",
[2020-11-18T23:05:46.376Z]   "LockAcquisitionTimeout": "10675199.02:48:05.4775807",
[2020-11-18T23:05:46.376Z]   "LockAcquisitionPollingInterval": "00:00:05",
[2020-11-18T23:05:46.376Z]   "ListenerLockRecoveryPollingInterval": "00:01:00"
[2020-11-18T23:05:46.376Z] }
[2020-11-18T23:05:46.376Z] HttpOptions
[2020-11-18T23:05:46.377Z] {
[2020-11-18T23:05:46.377Z]   "DynamicThrottlesEnabled": false,
[2020-11-18T23:05:46.377Z]   "MaxConcurrentRequests": -1,
[2020-11-18T23:05:46.377Z]   "MaxOutstandingRequests": -1,
[2020-11-18T23:05:46.377Z]   "RoutePrefix": "api"
[2020-11-18T23:05:46.377Z] }
[2020-11-18T23:05:46.380Z] Starting JobHost
[2020-11-18T23:05:46.388Z] Starting Host (HostId=84e8d03f575d-1810610370, InstanceId=cfff564f-dd78-4c11-91b5-c2cef0ba7f89, Version=3.0.14916.0, ProcessId=5253, AppDomainId=1, InDebugMode=False, InDiagnosticMode=False, FunctionsExtensionVersion=(null))
[2020-11-18T23:05:46.418Z] Loading functions metadata
[2020-11-18T23:05:46.419Z] FUNCTIONS_WORKER_RUNTIME set to node. Skipping WorkerConfig for language:java
[2020-11-18T23:05:46.420Z] FUNCTIONS_WORKER_RUNTIME set to node. Skipping WorkerConfig for language:powershell
[2020-11-18T23:05:46.421Z] FUNCTIONS_WORKER_RUNTIME set to node. Skipping WorkerConfig for language:python
[2020-11-18T23:05:46.436Z] Loading proxies metadata
[2020-11-18T23:05:46.443Z] Initializing Azure Function proxies
[2020-11-18T23:05:46.753Z] 0 proxies loaded
[2020-11-18T23:05:46.767Z] 10 functions loaded
[2020-11-18T23:05:46.854Z] Generating 10 job function(s)
[2020-11-18T23:05:46.946Z] Found the following functions:
  ---redacted---
[2020-11-18T23:05:46.949Z] 
[2020-11-18T23:05:46.971Z] Initializing function HTTP routes
  ---redacted---
[2020-11-18T23:05:46.975Z] 
[2020-11-18T23:05:46.989Z] Host initialized (585ms)
[2020-11-18T23:05:46.993Z] Host started (603ms)
[2020-11-18T23:05:46.993Z] Job host started

Functions:

  ---redacted---

[2020-11-18T23:05:47.289Z] Worker process started and initialized.
[2020-11-18T23:05:52.013Z] Host lock lease acquired by instance ID '00000000000000000000000006E61056'.
[2020-11-18T23:11:10.868Z] Worker process started and initialized.
[2020-11-18T23:17:40.906Z] Worker process started and initialized.
[2020-11-18T23:24:10.635Z] Exceeded language worker restart retry count for runtime:node. Shutting down and proactively recycling the Functions Host to recover
[2020-11-18T23:24:10.657Z] Stopping host...
[2020-11-18T23:24:10.679Z] Stopping JobHost
[2020-11-18T23:24:10.691Z] Job host stopped
[2020-11-18T23:24:10.717Z] Host shutdown completed.
```

**Comment by kpm-at-hfi** [comment]
I'm wondering if this might be related to developing in containers. I rolled back to just before I ran this in a container and it seems to run fine, without stopping.

I started running in a container by running "Remote Containers: Add Development Container Configuration Files..." and choosing "Azure Functions & Node.js". Which, BTW, was amazing for how seamlessly it got things running. The only flaw is this "Exceeded language worker restart..." stopping problem.

Any thoughts?
